### ISSUE TITLE
JMAP extensions: improve {from,to,cc,bcc}ContactGroupId filters to avoid unsupportedFilter: search too complex error

### ISSUE THREAD

**Issue by mildred** [cyrusimap/cyrus-imapd#4111]
I have a specific use case where

- I want to query all e-mail from/to a specific contact group (friends, coworkers, ...), and I also
- I want to query all e-mail that did not match a contact group, in order to assign the e-mails found to existing groups

For the first query, I'm using a OR operator to filter mail if any of the headers (From,To,Cc,Bcc) match a contact group, that's ok but it makes 4 filters in a OR condition, and if I add any more filters, I risk the unsupportedFilter with a search too complex.

To fix that it would be great to have a filter `anyContactGroupId` to match all four headers, and perhaps `recipientContactGroupId` to match to/cc/bcc.

For the second query, the search is definitely too complex as I chain in a NOT operator all the 4 headers filters for all the groups I have. If I have 5 groups, this produces 20 filters plus the NOT operator which definitely triggers the search too complex.

The same way the Email/query has a filter `inMailboxOtherThan`, it would be great to have a `{from,to,cc,bcc,any}ContactGroupOtherThan` taking a list of group ids.

Now, If I want to show the e-mails in threads, some more would be useful like the standard filters `allInThreadHaveKeyword`, `someInThreadHaveKeyword` and `noneInThreadHaveKeyword` with filters such as `allInThreadAnyInContactGroup`, `someInThreadAnyInContactGroup` and `noneInThreadAnyContactGroup`. I haven't yet investigated that but perhaps that could be instead be achieved via chaining requests.

**Comment by rsto** [comment]
BTW It is not an issue if you have a large OR query. But NOT(AND(OR))) queries with more than a couple of filter criteria can get a problem.

**Comment by mildred** [comment]
Thank you, I looked for but I couldn't find the code that implemented this limit. Could you point it to me so I can get a better idea on how it works to prevent complex queries ?

**Comment by rsto** [comment]
The DNF normalization code is located in these lines: https://github.com/cyrusimap/cyrus-imapd/blob/b114daeeb8b13771f2dbb52353c876f1c6e3ea06/imap/search_expr.c#L479-L923

The entry point is https://github.com/cyrusimap/cyrus-imapd/blob/b114daeeb8b13771f2dbb52353c876f1c6e3ea06/imap/search_expr.c#L896

and the function that checks the complexity limit:
https://github.com/cyrusimap/cyrus-imapd/blob/b114daeeb8b13771f2dbb52353c876f1c6e3ea06/imap/search_expr.c#L183

The complexity limit can be configured in imap.conf:
https://github.com/cyrusimap/cyrus-imapd/blob/b114daeeb8b13771f2dbb52353c876f1c6e3ea06/lib/imapoptions#L2319-L2323

where "resource bound" actually means the maximum number of allowed nodes in the search expression tree.

**Comment by mildred** [comment]
From an implementation point of view, it makes sense to be able to fold a query like

    fromContactGroup(friends) OR toContactGroup(friends) OR ccContactGroup(friends)

to something like

    contactGroup(friends, [from, to, cc])

As you check the contact group only once.

**Comment by mildred** [comment]
What I don't understand is that I'm getting a search too complex with only AND and NOT (which is is AND with negated predicates) and without a single OR operator.

My query looks like:

* AND
    * ```
      has header: List-Id
      inMailboxOtherThan: Junk mailbox
    * NOT
        * `fromAnyContact: true`
    * `notKeyword: $seen`
    * `notKeyword: $ignored`
    * NOT
        * `header List-Id: A`
        * `header List-Id: B`
        * `header List-Id: C`
        * `header List-Id: D`
        * `header List-Id: E`

There is no OR operator, the complexity is of 10 operations, where is the complexity in such a query?

**Comment by rsto** [comment]
The `fromAnyContact` internally gets converted to an `OR(all your contacts in here)`

**Comment by rsto** [comment]
No, the `inMailboxOtherThan` filter is optimized to evaluate as a single expression in the search tree (see `emailsearch_folders_value` in imap/jmap_mail.c). But the ORs in the `fromAnyContact` filter need to get "pulled" up on the top of the expression tree to be valid DNF, across the NOT and AND. The number of clauses grows exponentially.

**Comment by mildred** [comment]
Trying to debug a little bit... it seems fromAnyContact does not produce that much complexity after all. Instead the complexity of my request seems to come from filtering from the List-Id header.

With the following query:

```json
    ["Email/query",
      {
        "accountId":"XXX",
        "collapseThreads":true,
        "filter":{
          "operator":"AND",
          "conditions":[
            {"inMailboxOtherThan":["1ib27hhyeq7bq6b7d9lmdpoc"]},
            {"operator":"NOT","conditions":[
              {"fromAnyContact":true}
            ]},
            {"notKeyword":"$seen"},
            {"notKeyword":"$ignored"},
            {"operator":"NOT","conditions":[
              {"header":["List-Id","<A>"]},
              {"header":["List-Id","<B>"]},
              {"header":["List-Id","JSON Message Access Protocol <jmap.ietf.org>"]},
              {"header":["List-Id","<C>"]},
              {"header":["List-Id","<D>"]},
              {"header":["List-Id","<E>"]}
            ]
          }
        ]
      },
      "sort":[{"property":"receivedAt","isAscending":false}],
      "position":0,"limit":100
    },"query"],
```

I'm getting a nnodes of 1018

- removing a List-Id filter, I get a nnodes of 1008
- removing 2 list-Id filter, I get nnodes of 1003
- removing 3 list-id filters I get nnodes of 446
- removing 4 list-id filters (leaving 2) I get nnodes of 198
- removing 5 list-id filters (leaving 1) I get nnodes of 88
- removing all List-Id filters seems to get to a nnodes of 26
- removing NOT fromAnyContact: true from the original request I get a nnodes of 1008
- removing inMailboxOtherThan from the original request I get a nnodes of 1014
- removing both NOT fromAnyContact and inMailboxOtherThan I get nnodes of 1008
- removing both notKeywords filters I get a nnodes of 1002
- removing collapseThreads does not change anything

To get the nnodes, I gdb the httpd process and run
```
break complexity_check
commands
print *nnodes
c
end
```

**Comment by mildred** [comment]
I don't see why normalizing the filter expression could cause it to grow exponentially. And in this particular case, it seems there is not a single OR node after normalization.

Debugging the exponential development of the query it seems it starts to happens in `apply_distribution` called by `invert` called by `normalise`. Here, the normalization step is taking each child node (each child of the root AND node here). If the child node is the same operator as the parent node (same dnf_depth) it combines it with the parent node and restarts the normalization. In case the child node should appear higher, it calls invert, else it normalizes further.

The structure is in the middle of the normalization (nnodes = 35):

- OR
    - AND
        - TRUE
        - MATCH _emailsearch_folders_otherthan_attr
        - AND
            - TRUE
            - NOT
                - MATCH "indexflags"
        - AND
            - TRUE
            - NOT
                - MATCH "keyword"
        - NOT
            - OR
                -  AND
                    - TRUE
                    - MATCH header:list-id A
                - AND
                    - TRUE
                    - MATCH header:list-id B
                - AND
                    - TRUE
                    - MATCH header:list-id C
                - AND
                    - TRUE
                    - MATCH header:list-id D
                - AND
                    - TRUE
                    - MATCH header:list-id E
                - AND
                    - TRUE
                    - MATCH header:list-id F
        - NOT
            - TRUE

Next, I believe the code will join level 2 and level 3 AND operators. Then it will invert the OR and NOT with demorgan. it should come to a representation without too much OR nodes in the end. The problem is that it seems to be generating many nodes that are just plain junk (all the TRUE nodes).

There is also the NOT TRUE at the end that I don't understand. It should invalidate the whole query...

**Comment by mildred** [comment]
Perhaps the search query in itself is not that complex, however the process of normalizing it creates too many nodes. Perhaps detrivialise should be called often during the normalization process (if it can recycle nodes, else trivial nodes should be recycled to a pool of free nodes that can be used instead of allocating new nodes)

**Comment by mildred** [comment]
Perhaps a simple improvement for JMAP would be to change how the NOT operator is translated into a search query. Currently, a NOT(..., ...) operator is translated into NOT(OR(..., ...)). It could easily directly be translated into AND(NOT(...),NOT(...)). It is thus already in the correct order for the normalization and should improve a lot. In every case, the individual NOT operators are going to be created.

**Comment by rsto** [comment]
> Perhaps detrivialise should be called often during the normalization process

> Currently, a NOT(..., ...) operator is translated into NOT(OR(..., ...)). It could easily directly be translated into AND(NOT(...),NOT(...)).

These both are interesting points, thanks for bringing it up. I am currently busy with other stuff, but will have a look at it next 2 weeks.

**Comment by rjbs** [comment]
This problem is a known issue in our search, but I think this issue is not needed.  We know we want to fix the optimization of some issues like this in the future.  Closing this issue.
